tutorials/define_and_compile_model_webgl.py

```diff
@@ -0,0 +1,289 @@
+# to show the network structure.
+
+batch_size = 1
+num_class = 1000
+image_shape = (3, 224, 224)
+data_shape = (batch_size,) + image_shape
+out_shape = (batch_size, num_class)
+
+net, params = nnvm.testing.resnet.get_workload(
+    batch_size=batch_size, image_shape=image_shape)
+print(net.debug_str())
+
+######################################################################
+# Compilation
+# -----------
+# Next step is to compile the model using the NNVM/TVM pipeline.
+# Users can specify the optimization level of the compilation.
+# Currently this value can be 0 to 2, which corresponds to
+# "SimplifyInference", "OpFusion" and "PrecomputePrune" respectively.
+# In this example we set optimization level to be 0
+# and use OpenGL as compile target.
+#
+# :any:`nnvm.compiler.build` returns three components: the execution graph in
+# json format, the TVM module library of compiled functions specifically
+# for this graph on the target hardware, and the parameter blobs of
+# the model. During the compilation, NNVM does the graph-level
+# optimization while TVM does the tensor-level optimization, resulting
+# in an optimized runtime module for model serving.
+#
+# To generate the module library, TVM will first transfer graph IR into lower
+# intrinsic IR for the specified target backend, which is OpenGL in this
+# example. Then target backend will generate module library.
+
+if not tvm.module.enabled("opengl"):
+    print("OpenGL backend not enabled. This tutorial cannot be run.")
+else:
+    with nnvm.compiler.build_config(opt_level=0):
+        graph, lib, params = nnvm.compiler.build(
+            net,
+            target=tvm.target.opengl(),
+            shape={"data": data_shape},
+            params=params)
+
+######################################################################
+# Save Compiled Module
+# --------------------
+# After compilation, we can save the graph, lib and params into separate files
+# and deploy them to OpenGL.
+
+from tvm.contrib import util
+
+if not tvm.module.enabled("opengl"):
+    print("OpenGL backend not enabled. This tutorial cannot be run.")
+else:
+    temp = util.tempdir()
+
+    path_lib = temp.relpath("deploy_lib.so")
+    path_graph_json = temp.relpath("deploy_graph.json")
+    path_params = temp.relpath("deploy_param.params")
+
+    lib.export_library(path_lib)
+    with open(path_graph_json, "w") as fo:
+        fo.write(graph.json())
+    with open(path_params, "wb") as fo:
+        fo.write(nnvm.compiler.save_param_dict(params))
+
+    print(temp.listdir())
+
+######################################################################
+# Deploy locally to OpenGL
+# ------------------------
+# Now we can load the module back.
+
+import numpy as np
+from tvm.contrib import graph_runtime
+
+if not tvm.module.enabled("opengl"):
+    print("OpenGL backend not enabled. This tutorial cannot be run.")
+else:
+    loaded_lib = tvm.module.load(path_lib)
```

Comment:
Isolate deploy_local, deploy_rpc, deploy_web into three functions, and add proper comments

Reply:
Changed the 3 demos into separated functions.